pokemongo_bot/event_handlers/chat_handler.py

```diff
@@ -207,73 +67,131 @@ def get_caught(self, chat_id, num, order):
+                    "_Poke Stop Visits:_ " + str(stats["poke_stop_visits"]) + " (" + str(ps_day) + " _last 24h_)",
+                    "_KM Walked:_ " + str("%.2f" % stats["km_walked"])
+                )
+            return (res)
+        else:
+            return ("Stats not loaded yet\n")
+
+    def showtop(self, num, order):
+        if not num.isnumeric():
+            num = 10
+        else:
+            num = int(num)
 
-        def upgrade(self, chatid, uid):
-            # TODO: here comes upgrade logic (later)
-            self.sendMessage(chat_id=chatid, parse_mode='HTML', text="Upgrade logic not implemented yet")
-            return
\ No newline at end of file
+        if order not in ["cp", "iv"]:
+            order = "iv"
+        pkmns = sorted(inventory.pokemons().all(), key=lambda p: getattr(p, order), reverse=True)[:num]
+        return pkmns
+
+    def get_events(self, update):
+        cmd = update.message.text.split(" ", 1)
+        if len(cmd) > 1:
+            # we have a filter
+            event_filter = ".*{}-*".format(cmd[1])
+        else:
+            # no filter
+            event_filter = ".*"
+        events = filter(lambda k: re.match(event_filter, k), self.bot.event_manager._registered_events.keys())
+        events = sorted(events)
+        return events
+
+    def get_event(self, event, formatted_msg, data):
+        msg = None
+        if event == 'level_up':
+            msg = "level up ({})".format(data["current_level"])
+        elif event == 'pokemon_caught':
+            trigger = None
+            if data["pokemon"] in self.pokemons:
+                trigger = self.pokemons[data["pokemon"]]
+            elif "all" in self.pokemons:
+                trigger = self.pokemons["all"]
+            if trigger:
+                if ((not "operator" in trigger or trigger["operator"] == "and") and data["cp"] >= trigger["cp"] and
+                            data["iv"] >= trigger["iv"]) or \
+                        ("operator" in trigger and trigger["operator"] == "or" and (
+                                data["cp"] >= trigger["cp"] or data["iv"] >= trigger["iv"])):
+                    msg = "Caught {} CP: {}, IV: {}".format(data["pokemon"], data["cp"], data["iv"])
+        elif event == 'egg_hatched':
+            msg = "Egg hatched with a {} CP: {}, IV: {} (A/D/S {})".format(data["name"], data["cp"], data["iv_pct"],
+                                                                           data["iv_ads"])
+        elif event == 'bot_sleep':
+            msg = "I am too tired, I will take a sleep till {}.".format(data["wake"])
+        elif event == 'catch_limit':
+            msg = "*You have reached your daily catch limit, quitting.*"
+        elif event == 'spin_limit':
+            msg = "*You have reached your daily spin limit, quitting.*"
+        else:
+            return formatted_msg
+
+        return msg
+
+    def sendTeleMessage(self, chat_id=None, parse_mode='Markdown', text=None):
```

Comment:
Should be no sendmessage functionality in chathandler, only return data to caller. If this isn't being used, remove it, otherwise more work required.